### DISCUSSION TITLE
When converting to safetensor after full fine-tuning

### DISCUSSION THREAD

**Discussion by Foreist** [axolotl-ai-cloud/axolotl#3163] [Q&A] [answered]
The transformed safetensor cannot be read by transformers.
A torch.bin file and a reference file listing layer names for each bin are generated. How can I properly convert these files into a safetensor?

**Comment by NanoCode012** [comment]
Hello, could you provide more context behind your issue? Can you list that directory? Which commands did you use to generate that model?

By default, it should already be in safe tensors.

**Reply by Foreist** [reply]
`# This is the huggingface model that contains *.pt, *.safetensors, or *.bin files
# This can also be a relative path to a model on disk
base_model: /workspace/data/workspace/model/gemma-3-12b-it
# If you want to specify the type of model to load, AutoModelForCausalLM is a good choice too
model_type: AutoModelForCausalLM
# Corresponding tokenizer for the model AutoTokenizer is a good choice
tokenizer_type: AutoTokenizer
# Trust remote code for untrusted source
trust_remote_code: True

# Resize the model embeddings when new tokens are added to multiples of 32
# This is reported to improve training speed on some models
resize_token_embeddings_to_32x: True


bnb_config_kwargs:
  # These are default values
  llm_int8_has_fp16_weight: false
  bnb_4bit_quant_type: nf4
  bnb_4bit_use_double_quant: false

bf16: true # bool or 'full' for `bf16_full_eval`. require >=ampere
# Use CUDA fp16
fp16: false
# Use CUDA tf32
tf32: false # require >=ampere

# No AMP (automatic mixed precision) amp를 위 옵션처럼 사용하는 거니까 false로 사용하게 하기
bfloat16: true # require >=ampere
float16: false

# A list of one or more datasets to finetune the model with
datasets:
  # HuggingFace dataset repo | s3://,gs:// path | "json" for local dataset, make sure to fill data_files
  - path: /workspace/data/workspace/data/250911_no_shuffle-10task-train_data.jsonl
    # The type of prompt to use for training. [alpaca, gpteacher, oasst, reflection]
    chat_template: tokenizer_default
    type: chat_template # format | format:<prompt_style> (chat/instruct) | <prompt_strategies>.load_<load_fn>
    field_messages: conversations
    ds_type: # Optional[str] (json|arrow|parquet|text|csv) defines the datatype when path is a file
    data_files: # Optional[str] path to source data files
    shards: # Optional[int] number of shards to split data into
    name: # Optional[str] name of dataset configuration to load
    train_on_split: # Optional[str] name of dataset split to load from
    revision: # Optional[str] The specific revision of the dataset to use when loading from the Hugging Face Hub. This can be a commit hash, tag, or branch name. If not specified, the latest version will be used. This parameter is ignored for local datasets.

    # Optional[List[str]]. Roles to train on. The tokens from these roles will be considered for the loss.
    roles_to_train: ["assistant"]
    # Optional[str]. Which EOS tokens to train on in the conversation. Possible values are:
    # - all: train on all EOS tokens
    # - turn: train on the EOS token at the end of each trainable turn
    # - last: train on the last EOS token in the conversation
    train_on_eos: turn

special_tokens:
  eos_token: '<eos>'
eot_tokens:
  - "<end_of_turn>"

# A list of one or more datasets to eval the model with.w
# You can use either test_datasets, or val_set_size, but not both.
test_datasets:
  - path: /workspace/data/workspace/data/250911_no_shuffle-10task-valid_data.jsonl
    # ds_type: jsonl
    chat_template: tokenizer_default
    # You need to specify a split. For "json" datasets the default split is called "train".
    split: train
    type: chat_template # format | format:<prompt_style> (chat/instruct) | <prompt_strategies>.load_<load_fn>
    field_messages: conversations

# If false, the datasets will not be shuffled and will keep their original order in `datasets`.
# The same applies to the `test_datasets` option and the `pretraining_dataset` option. Default is true.
shuffle_merged_datasets: true
# The name of the chat template to use for training, following values are supported:
# - tokenizer_default: Uses the chat template that is available in the tokenizer_config.json. If the chat template is not available in the tokenizer, it will raise an error. This is the default value.
# - alpaca/inst/chatml/gemma/cohere/llama3/phi_3/deepseek_v2/jamba: These chat templates are available in the axolotl codebase at src/axolotl/utils/chat_templates.py
# - tokenizer_default_fallback_*: where * is the name of the chat template to fallback to. E.g. tokenizer_default_fallback_chatml. This is useful when the chat template is not available in the tokenizer.
# - jinja: Uses a custom jinja template for the chat template. The custom jinja template should be provided in the chat_template_jinja field.
# The selected chat template will be saved to the tokenizer_config.json for easier inferencing
# Note: It is recommended to set train_on_inputs to true when using a chat template that is different from the model's default chat template.
chat_template: tokenizer_default


# Axolotl attempts to save the dataset as an arrow after packing the data together so
# subsequent training attempts load faster, relative path
dataset_prepared_path: /workspace/last_run_prepared

# How much of the dataset to set aside as evaluation. 1 = 100%, 0.50 = 50%, etc. 0 for no eval.
# val_set_size: 0.01
# Num shards for whole dataset
dataset_shard_num:
# Index of shard to use for whole dataset
dataset_shard_idx:

# The maximum length of an input to train with, this should typically be less than 2048
# as most models have a token/context limit of 2048
sequence_len: 12500
# Pad inputs so each step uses constant sized buffers
# This will reduce memory fragmentation and may prevent OOMs, by re-using memory more efficiently
# OOM 날 리 없으니 ㄴㄴ
pad_to_sequence_len: true
# Use efficient multi-packing with block diagonal attention and per sequence position_ids. Recommend set to 'true'
#   [샘플1] [PAD] [PAD] [PAD]  # 패킹 전
#   [샘플2] [PAD] [PAD]
   
#   [샘플1][샘플2][샘플3]      # 패킹 후
sample_packing: true
# Set to 'false' if getting errors during eval with sample_packing on.
eval_sample_packing: false

# Increasing the following values helps with packing, but usually only slightly (<%1.)
# The number of samples packed at a time.
sample_packing_group_size: 100000
# The number of samples which can be packed into one sequence. Increase if using a large sequence_len with many short samples.
sample_packing_bin_size: 200


lora_fan_in_fan_out: false

# LoRA+ hyperparameters
# For more details about the following options, see:
# https://arxiv.org/abs/2402.12354  and `src/axolotl/core/train_builder.py`
# loraplus_lr_ratio: 16 # loraplus learning rate ratio lr_B / lr_A. Recommended value is 2^4.
# loraplus_lr_embedding: 1e-6 #  loraplus learning rate for lora embedding layers. Default value is 1e-6.



# Where to save the full-finetuned model to
output_dir: /workspace/data/workspace/model/axolotl/gemma3-all-2200_200-lora64_64-filter-pfs-1e-4-adapter1/
learning_rate: 1e-4
# Whether to use torch.compile and which backend to use
# torch_compile: true # bool
# torch_compile_backend: "inductor" # Optional[str]

# Training hyperparameters
lr_scheduler: "constant"

# If greater than 1, backpropagation will be skipped and the gradients will be accumulated for the given number of steps.
gradient_accumulation_steps: 1
# The number of samples to include in each batch. This is the number of samples sent to each GPU.
# Batch size per gpu = micro_batch_size * gradient_accumulation_steps
# 1280
micro_batch_size: 2
eval_batch_size: 2
num_epochs: 1
warmup_ratio: 0  # cannot use with warmup_steps
lr_quadratic_warmup:
eval_strategy: "steps"
logging_steps: 1
eval_steps:  200 # Leave empty to eval at each epoch, integers for every N steps. decimal for fraction of total steps
save_strategy: "epoch" # Set to `"no"` to skip checkpoint saves
save_steps: # Leave empty to save at each epoch
save_total_limit: 1 # Checkpoints saved at a time

eval_table_size: # Approximate number of predictions sent to wandb depending on batch size. Enabled above 0. Default is 0
eval_max_new_tokens: # Total number of tokens generated for predictions sent to wandb. Default is 128
# do_causal_lm_eval: true
# eval_causal_lm_metrics: ["perplexity"] # HF evaluate metrics used during evaluation. Default is ["sacrebleu", "comet", "ter", "chrf", "perplexity"]

loss_watchdog_threshold: # High loss value, indicating the learning has broken down (a good estimate is ~2 times the loss at the start of training)
loss_watchdog_patience: # Number of high-loss steps in a row before the trainer aborts (default: 3)

# Save model as safetensors (require safetensors package)
save_safetensors:

# Whether to mask out or include the human's prompt from the training labels
train_on_inputs: false
# Group similarly sized data to minimize padding.
# May be slower to start, as it must download and sort the entire dataset.
# Note that training loss may have an oscillating pattern with this enabled.
group_by_length: false

# Whether to use gradient checkpointing https://huggingface.co/docs/transformers/v4.18.0/en/performance#gradient-checkpointing
gradient_checkpointing: true
# additional kwargs to pass to the trainer for gradient checkpointing
# gradient_checkpointing_kwargs:
#   use_reentrant: true

# Stop training after this many evaluation losses have increased in a row
# https://huggingface.co/transformers/v4.2.2/_modules/transformers/trainer_callback.html#EarlyStoppingCallback
# early_stopping_patience: 10


# Specify optimizer
# Valid values are driven by the Transformers OptimizerNames class, see:
# https://github.com/huggingface/transformers/blob/95b374952dc27d8511541d6f5a4e22c9ec11fb24/src/transformers/training_args.py#L134
#
# Note that not all optimizers may be available in your environment, ex: 'adamw_anyprecision' is part of
# torchdistx, 'adamw_bnb_8bit' is part of bnb.optim.Adam8bit, etc. When in doubt, it is recommended to start with the optimizer used
# in the examples/ for your model and fine-tuning use case.
#
# Valid values for 'optimizer' include:
# - adamw_hf
# - adamw_torch
# - adamw_torch_fused
# - adamw_torch_xla
# - adamw_apex_fused
# - adopt_adamw (only for torch version >= 2.5.1)
# - adafactor
# - adamw_anyprecision
# - sgd
# - adagrad
# - adamw_bnb_8bit
# - lion_8bit
# - lion_32bit
# - paged_adamw_32bit
# - paged_adamw_8bit
# - paged_lion_32bit
# - paged_lion_8bit
# - galore_adamw
# - galore_adamw_8bit
# - galore_adafactor
# - galore_adamw_layerwise
# - galore_adamw_8bit_layerwise
# - galore_adafactor_layerwise
optimizer: 'adamw_torch_fused'

# Whether to bettertransformers
flash_optimum: false
# Whether to use xformers attention patch https://github.com/facebookresearch/xformers:
xformers_attention: false
# Whether to use flash attention patch https://github.com/Dao-AILab/flash-attention:
# gemma2 train 시에 flash_attention2 쓰지 말라고 함 ㅠ
flash_attention: true
flash_attn_cross_entropy: false # Whether to use flash-attention cross entropy implementation - advanced use only
flash_attn_rms_norm: true # Whether to use flash-attention rms norm implementation - advanced use only
flash_attn_fuse_qkv: true # Whether to fuse QKV into a single operation
flash_attn_fuse_mlp: true # Whether to fuse part of the MLP into a single operation
# Whether to use scaled-dot-product attention
# https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
sdp_attention: false
# Shifted-sparse attention (only llama) - https://arxiv.org/pdf/2309.12307.pdf
s2_attention: false
# Resume from a specific checkpoint dir
resume_from_checkpoint:
# If resume_from_checkpoint isn't set and you simply want it to start where it left off.
# Be careful with this being turned on between different models.
auto_resume_from_checkpoints: false

# Don't mess with this, it's here for accelerate and torchrun
local_rank:

# Deepspeed config path. e.g., deepspeed_configs/zero3.json
deepspeed: /workspace/axolotl/deepspeed_configs/zero3_bf16.json

# liger
plugins:
  - axolotl.integrations.liger.LigerPlugin
  # - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_layer_norm: true
liger_fused_linear_cross_entropy: true
# chunked_cross_entropy: false

# Optional[bool] Don't upcast the embeddings to float32 when using PEFT. Useful for low-VRAM GPUs(정말 메모리 없을 때만)
embeddings_skip_upcast: false

dataloader_pin_memory: True

# Debug mode
debug: false

# Seed
seed: 1234`

there's my config. 
"By default, it should already be in safe tensors." <- really? T^T

**Reply by NanoCode012** [reply, marked as answer]
Hey, it's because `save_safetensors:` is empty.

You're also using quite an old config example. I would recommend starting from the ones in our example directory as your new base https://github.com/axolotl-ai-cloud/axolotl/blob/main/examples/gemma3/gemma-3-4b-qlora.yml

The above is a 4B example, you just need to replace base path to your 12b

**Reply by Foreist** [reply]
> Hey, it's because `save_safetensors:` is empty.
> 
> You're also using quite an old config example. I would recommend starting from the ones in our example directory as your new base https://github.com/axolotl-ai-cloud/axolotl/blob/main/examples/gemma3/gemma-3-4b-qlora.yml
> 
> The above is a 4B example, you just need to replace base path to your 12b

I love you.
